### DISCUSSION TITLE
CAN bus for ESP32 or RP2040?

### DISCUSSION THREAD

**Discussion by ilium007** [micropython/micropython#9584] [Libraries & Drivers]
Trying to search for "CAN" in the old forums is pretty much impossible! I found a few threads were the discussion of CANbus libraries for boards other than the Pyboard came up but some of these threads are 6 years old. Does anyone know of work being done to get a CAN library working on the ESP32 or RP2040 boards? I have been successful with CANbus on STM32F405 dev boards but would like to do some prototyping on the ESP32 (due to network access) but need working CANbus for my design.

**Comment by ilium007** [comment]
This is as close as I have found by searching - [https://github.com/micropython/micropython/pull/9532](https://github.com/micropython/micropython/pull/9532)

**Comment by obdevel** [comment]
As you no doubt know, the RP2040 doesn't have a CAN peripheral but it's quite straightforward to get an external MCP2515 chip (and family members) working. Adafruit has a CircuitPython library and I have been doing some work to port an Arduino C++ library across. It's just reading and writing registers over SPI, as documented in the datasheet. The problem is finding a module that works at 3.3V, if you can't make your own. I have a Pico 'hat' PCB design, if you can do surface mount assembly.

**Reply by ilium007** [reply]
Do you have a link to your pico hat? Just looked at your MergCBus repo... I project is canbus control of model railways!!!!

**Reply by obdevel** [reply]
MERG (merg.org.uk) is a group that uses CAN as a layout control bus for model railways (called CBUS). The project you mention is a port of my Arduino C/C++ library to MicroPython. The group's repo is: https://github.com/MERG-DEV

The board has additional components for the switch and LEDs that CBUS needs. I haven't published the design files yet as I'm still testing and I may make some changes. I'm also designing a through-hole version but it's larger.

![IMG_0171](https://user-images.githubusercontent.com/13898399/196451958-dcd4ed4d-3cda-47fa-8a66-5ce63fa12a4f.JPG)

**Reply by ilium007** [reply]
> As you no doubt know, the RP2040 doesn't have a CAN peripheral but it's quite straightforward to get an external MCP2515 chip (and family members) working. Adafruit has a CircuitPython library and I have been doing some work to port an Arduino C++ library across. It's just reading and writing registers over SPI, as documented in the datasheet. The problem is finding a module that works at 3.3V, if you can't make your own. I have a Pico 'hat' PCB design, if you can do surface mount assembly.

I can make my own 3.3v controller / transceiver hardware, I just haven't found an appropriate library. The MCP2515 is listed as "Not Recommended for new designs" and is replaced by the MCP2518FD. I'm looking at the RP2040 as the STM32F405 is too hard to source at the moment, the RP2040 is readily available. I'm surprised there aren't more people using these for CANbus projects.

**Comment by ilium007** [comment]
A company, CANIS, has exactly what I'm after - https://kentindell.github.io/2021/03/10/micropython-can-api

Not sure if they have been able to release source yet but they are using the newer MCP2518FD with CAN ID filters.

**Reply by RufusVS** [reply]
It's unfortunate that this wasn't asked as two separate questions.  RP2040 doesn't have an internal CAN controller, though it does have those special engines that might be programmable for it.   But the ESP32 DOES have a CAN controller built in, so a python CAN module is conspicuous in its absence.  There must be some code portable into micropython somewhere.

I'll probably enter it as a separate topic and link to it.

**Reply by obdevel** [reply]
There is an existing project (in C and Arm assembler) that implements a CAN 2.0B controller in an RP2040 PIO block (search: can2040). It may be possible for someone to integrate this into the MP interpreter. It works surprisingly well.

Similarly, the ESP32 CAN controller peripheral is well-supported and documented by the IDF (Espressif call it TWAI - two-wire automotive interface, presumably so they don't have to pay Bosch the royalties !)

I have published a couple of MCP2515-based CAN 'shields' for the Pico. They're thru-hole designs so anyone could assemble them. The Gerbers are in my repo. There's a decent driver linked at awesome-micropython.com.
